hal/src/rtl872x/spi_hal.cpp

```diff
@@ -216,47 +224,64 @@ class Spi {
         if (spiConfig && spiConfig->flags) {
             config_.flags = spiConfig->flags;
         }
+        // Set data mode
+        setDataMode(config_.dataMode, force);
+        // Set clock divider
+        setClockDivider(config_.clockDiv);
+        // Set bit order
+        setBitOrder(config_.bitOrder);
+        return SYSTEM_ERROR_NONE;
+    }
 
-        uint32_t phase = SCPH_TOGGLES_IN_MIDDLE;
-        uint32_t polarity = SCPOL_INACTIVE_IS_LOW;
-        spiModeToPolAndPha(config_.dataMode, &polarity, &phase);
-        uint32_t bakPol = SPI_DEV_TABLE[rtlSpiIndex_].SPIx->CTRLR0 & BIT_CTRLR0_SCPH;
-        SSI_SetSclkPhase(SPI_DEV_TABLE[rtlSpiIndex_].SPIx, phase);
-        SSI_SetSclkPolarity(SPI_DEV_TABLE[rtlSpiIndex_].SPIx, polarity);
-        SSI_SetDataFrameSize(SPI_DEV_TABLE[rtlSpiIndex_].SPIx, DEFAULT_DATA_BITS);
-        if (force || (bakPol != (SPI_DEV_TABLE[rtlSpiIndex_].SPIx->CTRLR0 & BIT_CTRLR0_SCPH))) {
-            // PAD_PullCtrl is another slow call, avoid if settings still match
-            if (polarity == SCPOL_INACTIVE_IS_LOW) {
-                PAD_PullCtrl((uint8_t)hal_pin_to_rtl_pin(sclkPin_), GPIO_PuPd_DOWN);
+    int setBitOrder(uint8_t bitOrder) {
+        config_.bitOrder = bitOrder;
+        if (SPI_DEV_TABLE[rtlSpiIndex_].SPIx->SSIENR & BIT_SSIENR_SSI_EN) {
+            if (config_.bitOrder == MSBFIRST) {
+                SSI_SetDataSwap(SPI_DEV_TABLE[rtlSpiIndex_].SPIx, BIT_CTRLR0_TXBITSWP, DISABLE);
+                SSI_SetDataSwap(SPI_DEV_TABLE[rtlSpiIndex_].SPIx, BIT_CTRLR0_RXBITSWP, DISABLE);
             } else {
-                PAD_PullCtrl((uint8_t)hal_pin_to_rtl_pin(sclkPin_), GPIO_PuPd_UP);
+                SSI_SetDataSwap(SPI_DEV_TABLE[rtlSpiIndex_].SPIx, BIT_CTRLR0_TXBITSWP, ENABLE);
+                SSI_SetDataSwap(SPI_DEV_TABLE[rtlSpiIndex_].SPIx, BIT_CTRLR0_RXBITSWP, ENABLE);
             }
         }
+        return SYSTEM_ERROR_NONE;
+    }
 
-        // Set clock divider
-        if (config_.spiMode == SPI_MODE_MASTER) {
+    int setDataMode(uint8_t dataMode, bool force = false) {
+        config_.dataMode = dataMode;
+        if (SPI_DEV_TABLE[rtlSpiIndex_].SPIx->SSIENR & BIT_SSIENR_SSI_EN) {
```

Comment:
This is a new check to see if the SPIx peripheral is enabled before configuring it? Was there a problem before when configuring data mode/clock/bit order without checking this?

Reply:
The intention was to make sure the `setDataMode()` (clock/bit_order) just set internal variables if `begin()` is not called yet. Also the SDK APIs demonstrate we should disable the peripheral before setting these configurations.